### DISCUSSION TITLE
Proposal: Investigation of Matching Methods satisfying the Equal Percentage Bias Reduction (EPBR) criterion

### DISCUSSION THREAD

**Discussion by WenyiWU0111** [stoat-proj/SToAT#1] [Project Proposal]
Proposal File: [Project_Proposal_Matching.pdf](https://github.com/user-attachments/files/15845695/Project_Proposal_Matching.pdf)

Contributor: Wenyi WU
Date: June 11, 2024

## Introduction and Background

In the causal inference field, accurately estimating the treatment effect from observational data is a focus of study. In randomized experiments, the treatment effect can be estimated by taking the average difference between the treatment and control groups, as the two groups are considered homogeneous and comparable. However, in observational studies, subjects in treatment and control groups often differ due to confounding factors, and the observed differences in outcomes may not reflect the true treatment effect [^Benedetto2018]. Therefore, researchers aim to estimate the counterfactual outcome for each subject, i.e., the outcome that would have been observed if the subject had been in the opposite treatment group.

Matching is a classic and widely used method to estimate the treatment effect in observational data. It can be seen as estimating the counterfactual outcome of a subject in the treatment (control) group with the factual outcome of a similar subject in the control (treatment) group. For example, Mahalanobis Distance Matching measures the similarity of subjects in two groups with the Mahalanobis distance on the covariates $X$, but it suffers from the curse of dimensionality. Propensity Score Matching (PSM) addresses this issue by matching on the propensity score $p(t|X)$, which is the probability of being treated given the covariates $X$ [^rosenbaum1983]. However, PSM must satisfy the Equal Percentage Bias Reduction (EPBR) criterion to obtain an unbiased treatment effect estimation [^rubin1976]. Coarsened Exact Matching (CEM) satisfies the EPBR criterion, but when the dimension of covariates $X$ is high, only a few samples can be matched [^iacus2012].

Therefore, we want to investigate matching methods that satisfy the EPBR criterion and are capable of handling high-dimensional data.

## Settings

The following settings follow the Rubin Causal Model.

A causal effect of a subject $i$ can be defined as the difference
between an observed outcome $Y_{i1}$, when subject $i$ is treated, and its counterfactual $Y_{i0}$, when subject $i$ is untreated. Here we denote covariates of subject $i$ as $X_i$, and the treatment indicator as $T_i$. The treatment effect can then be estimated as 

$$
\tau = \hat{f_1}(X, T_i=1) - \hat{f_0}(X, T_i=0)
$$

where $f_1$ and $f_0$ are functions estimating the outcome $Y$. For example, the average treatment effect ATE can be seen as a special case.

$$
\tau = E(Y_{i1}|T_i=1) - E(Y_{i0}|T_i=0)
$$

However, the key challenge is that we cannot observe both $Y_{i1}$ and $Y_{i0}$ for the same subject, as a subject can only be either treated or untreated. Then we want to estimate the counterfactual outcome of subject $i$ using subjects who are similar to $i$ in the counterfactual group. This is where matching methods, such as PSM and CEM, come into play.

Therefore, we want to find some good matching methods $M(X_i)$ to match subject $i$ with similar subjects in the counterfactual group. Here we define good as:

1. Equal Percent Bias Reducing (EPBR): The matching method should reduce the bias equally across all covariates. (More details about EPBR can be checked in the Appendix)
2. Passing the Balance Checking: The matching method should achieve covariate balance, as assessed by the balance checking methods described [here](https://cem-linearinf.readthedocs.io/en/latest/tuto_balance.html).
3. Capability for high-dimensional data: The matching method should be able to handle high-dimensional covariate spaces.

We will compare the matching performance of these methods by doing experiments on simulation and empirical datasets that satisfy or violate the EPBR assumption, with both low and high dimensions.

Moreover, a non-EPBR matching method could increase the bias of our estimation if the outcome Y is linearly dependent with $X$. However, this is not always the case and there have been many state-of-art nonlinear estimation functions these days like Double Machine Learning [^Chernozhukov2018], TARNET [^shalit2017], and so on. Therefore, we also want to investigate whether estimations of these state-of-the-art nonlinear methods are still vulnerable to the EPBR criterion.

## Aims

1. Find matching methods that satisfy the EPBR criterion and implement them in a Python Package.
2. Evaluate the performance of these methods by doing experiments on both simulated and empirical data, considering the following factors:
    - Low and high-dimensional covariate spaces
    - Datasets that satisfy or violate the EPBR assumption
3. Estimate Individual Treatment Effects ($ITE$) and Average Treatment Effects ($ATE$) using the matched data, employing both classical and state-of-the-art deep learning methods. This will investigate how the quality of the matching process affects the subsequent treatment effect estimation.

## Related Works

There have been some matching algorithms and corresponding software packages that are useful for satisfying the EPBR criterion, and some works compare the performance of different matching methods or treatment effect estimation algorithms. The following is a summary of related works and existing gaps. Distinct features and significance of our project are also discussed.

### Works related to Aim 1
As stated above, matching methods that satisfy the EPBR criterion and are suitable for high-dimensional covariates $X$ are needed. Existing methods can be divided into three categories.

1. Methods that satisfy EPBR criterion naturally, such as CEM which matches samples with exactly the same coarsened covariates together. 
2. Methods that add some moment conditions while optimizing the distance between matched samples, such as Covariate Balance Propensity Score (CBPS) [^imai2014]. 
3. Methods that optimize the imbalance measure, i.e. Standardized Mean Difference, and the distance between matched samples simultaneously, such as Genetic Matching [^diamond2013].

The following table shows the corresponding packages of matching methods mentioned above, however, there are some problems with the existing packages:

1. These packages are primarily available in R, but a Python version is also needed to cater to a wider audience, which will be discussed in the following part.
2. Some of them are not suitable for high dimensional covariates.
3. Some of them do not provide treatment effect estimation like average treatment effect on treated $ATT$ and $ATE$ mentioned before. 
Knowing $ATT$ and $ATE$ can provide some foresee insights for the following heterogeneous treatment effect estimation with advanced methods. Although we can calculate these estimations manually or using other packages after getting matched pairs, it will be more convenient to integrate them within the matching package itself.

|                                         Methods                                        |   Package   | Language |                                            Functions                                           | Suitable for High Dimensional Covariates | Treatment Effect Estimation |
|:--------------------------------------------------------------------------------------:|:-----------:|:--------:|:----------------------------------------------------------------------------------------------:|:----------------------------------------:|:---------------------------:|
|                            Satisfy EPBR criterion naturally                            |   [Matchit](https://kosukeimai.github.io/MatchIt/)   |     R    |                                               CEM                                              |                    No                    |             Yes             |
|               Adds moment conditions while optimizing the distance               |     [CBPS](https://cran.r-project.org/web/packages/CBPS/)    |     R    |                                              CBPS                                              |                    Yes                   |             Yes             |
|               Adds moment conditions while optimizing the distance               | [Designmatch](https://cran.r-project.org/web/packages/designmatch/index.html) |     R    | Allow directly balancing moments and distributions of $X$ |                    Yes                   |              No             |
| optimize the imbalance measure and the distance between matched samples simultaneously |   [Matching](https://github.com/cran/Matching)  |     R    |                                        Genetic Matching                                        |                    No                    |              No             |

#### Why do we need the Python version

Many Matching algorithms have excellent R version Packages but no Python version because of the dominance of R language in the Statistics field. However, the increasing prominence of deep learning-related methods in the domain of causal inference, coupled with the computational efficiency of Python, has resulted in the emergence of advanced treatment effect estimation packages in Python, such as EconML [^econml], CausalML [^causalml], etc. Given that matching can serve as a crucial preprocessing step for treatment effect estimation, it is imperative to have a Python version of the excellent matching algorithms that are currently available primarily in R.

### Works related to Aim 2 & 3

In many matching algorithm papers, authors compare the performance of their proposed methods with other existing ones. For example, Imai and Ratkovic tested the covariates balancing performance of the Covariate Balance Propensity Score CBPS on the simulated dataset and compared it with PSM [^imai2014], but the simulated dataset only has 4 covariates and each covariate follows a normal distribution, which makes it satisfy the EPBR assumption naturally. Diamond and Sekhon compared the treatment effect estimation ability of Genetic Matching, Mahalanobis Distance Matching, and PSM on the NSW dataset [^diamond2013]. This dataset can test the matching methods’ ability to address the EPBR problem, but they did not check the covariates balancing performance of these methods. Rubin, etc. compared CEM, MDM, and PSM’s covariates balancing performance on four datasets [^king2011], but these datasets are not high dimensional.

Obviously, several notable gaps warrant further investigation:

1. Limited Evaluation of High-Dimensional Datasets 
2. Lack of Evaluation on Datasets that Violate EPBR Assumptions 
3. Incomprehensive Comparative Assessments

There are also some benchmarks comparing the treatment effect estimation performance of existing estimation algorithms, and some papers compare their proposed algorithms with others, such as [IHDP](https://paperswithcode.com/sota/causal-inference-on-ihdp) benchmark and [Jobs](https://paperswithcode.com/sota/causal-inference-on-jobs) benchmark provided in the Papers With Code, and the result of Methods Comparison for counterfactual inference in Schwab etc.’s paper [^schwab2018].
However, these results focus on evaluating inference algorithms themselves, without explicitly incorporating matching as a preprocessing step, and little attention is paid to the impact of matching quality on the performance of subsequent treatment effect estimation.

Therefore, Aim 2 and Aim 3 are meaningful for investigation.

## Timeline

1. Literature Review to find suitable matching methods that satisfy the EPBR criterion and capable of handling high-dimensional covariates $X$. [~ 0.5 month]

2. Implement selected methods in the Python package (if there is no existing one) and test it. [~ 2 month]

3. Simulate datasets and find empirical datasets that satisfy or violate the EPBR assumption, with both low and high dimensions. [~ 0.5 months]

4. Evaluate the performance of different matching methods by doing experiments on simulated and empirical datasets. Summarize the characteristics of each matching method. [~ 1 month]

5. Estimate Individual Treatment Effects (ITE) and Average Treatment Effects (ATE) using the matched data, employing Difference in Mean, DML, and TARNET to investigate how the quality of the matching process affects the subsequent treatment effect estimation. [~ 1 month]

## References
[^Benedetto2018]: Benedetto, U., Head, S. J., Angelini, G. D., & Blackstone, E. H. (2018). Statistical primer: Propensity score matching and its alternatives. European Journal of Cardio-Thoracic Surgery, 53(6), 1112-1117. 

[^causalml]: Huigang Chen, Totte Harinen, Jeong-Yoon Lee, Mike Yung, and Zhenyu Zhao. Causalml: Python package for causal machine learning, 2020.

[^Chernozhukov2018]: Chernozhukov, V., Chetverikov, D., Demirer, M., Duflo, E., Hansen, C., Newey, W., & Robins, J. (2018). Double/debiased machine learning for treatment and structural parameters. The Econometrics Journal, 21(1), C1-C68. 

[^diamond2013]: Alexis Diamond and Jasjeet S Sekhon. Genetic matching for estimating causal effects: A general multivariate matching method for achieving balance in observational studies. Review of Economics and Statistics, 95(3):932–945, 2013.

[^iacus2012]: Iacus, S. M., King, G., & Porro, G. (2012). Causal Inference Without Balance Checking: Coarsened Exact Matching. Political Analysis, 20(1), 1–24. 

[^imai2014]: Kosuke Imai and Marc Ratkovic. Covariate balancing propensity score. Journal of the Royal Statistical Society Series B: Statistical Methodology, 76(1):243–263, 2014.

[^econml]: Maggie Hei Greg Lewis Paul Oka Miruna Oprescu Vasilis Syrgkanis Keith Battocchi, Eleanor Dillon. EconML: A Python Package for ML-Based Heterogeneous Treatment Effects Estimation. https://github.com/py-why/EconML, 2019. Version 0.15.1.

[^king2011]: Gary King, Richard Nielsen, Carter Coberley, James E Pope, and Aaron Wells. Comparative effectiveness of matching methods for causal inference. Unpublished manuscript, Institute for Quantitative Social Science, Harvard University, Cambridge, MA, 2011.

[^rosenbaum1983]: Rosenbaum, P. R., & Rubin, D. B. (1983). The central role of the propensity score in observational studies for causal effects. Biometrika, 70(1), 41-55.

[^rubin1976]: D. B. Rubin. Multivariate matching methods that are equal percent bias reducing, i: Some examples. Biometrics, 32(1):109–120, 1976.

[^schwab2018]: Patrick Schwab, Lorenz Linhardt, and Walter Karlen. Perfect match: A simple method for learning representations for counterfactual inference with neural networks. arXiv preprint arXiv:1810.00656, 2018.

[^shalit2017]: Shalit, U., Johansson, F. D., & Sontag, D. (2017). Estimating individual treatment effect: Generalization bounds and algorithms. In International Conference on Machine Learning.

**Comment by statmlben** [comment]
Wenyi, thank you for submitting your proposal, which looks excellent! Could you please confirm whether similar modules or libraries exist on GitHub or PyPI that offer comparable functionality? Additionally, I would appreciate it if you could elaborate on the discussion surrounding your proposal and highlight its distinct features in relation to existing works:)

(PS: if possible, pls directly use markdown in github)

**Reply by WenyiWU0111** [reply]
Hi Professor DAI, thank you for your comments! I revised the proposal accordingly, markdown version in github was also edited. I searched related existing works and libraries and discussed distinct features and significance of this project. Contents are added to the 'Related Works' section. You can have a look when you have time.

**Reply by statmlben** [reply]
Wenyi, thank you for your updates! After reviewing your proposal, I have decided to fund this project. However, as we discussed, I feel that the aims are too intensive, so perhaps we can first focus on a smaller project, such as developing a `Python interface for MatchIt` (https://kosukeimai.github.io/MatchIt/). Would you please update the proposal a little bit, then push to the main page. Thx!

**Reply by WenyiWU0111** [reply]
Hi Professor DAI, I wrote a revised proposal and sent a pull request for the Markdown file, and I'll update the table on the main page once the proposal is committed.
